langchain/src/vectorstores/googlevertexai.ts

```diff
@@ -0,0 +1,682 @@
+          namespace: key,
+          [listType]: valArray,
+        };
+
+        // Add it to the restriction list
+        restrictions.push(restriction);
+      }
+    }
+    return restrictions;
+  }
+
+  /**
+   * Create an index datapoint for the vector and document id.
+   * If an id does not exist, create it and set the document to its value.
+   * @param vector
+   * @param document
+   */
+  buildDatapoint(vector: number[], document: IdDocument): IndexDatapoint {
+    if (!document.id) {
+      // eslint-disable-next-line no-param-reassign
+      document.id = uuid.v4();
+    }
+    const ret: IndexDatapoint = {
+      datapointId: document.id,
+      featureVector: vector,
+    };
+    const restrictions = this.metadataToRestrictions(document.metadata);
+    if (restrictions?.length > 0) {
+      ret.restricts = restrictions;
+    }
+    return ret;
+  }
+
+  async delete(params: MatchingEngineDeleteParams): Promise<void> {
+    const options = {};
+    await this.removeDatapointClient.request(params.ids, options);
+  }
+
+  async similaritySearchVectorWithScore(
+    query: number[],
+    k: number,
+    filter?: this["FilterType"]
+  ): Promise<[Document, number][]> {
+    // Format the query into the request
+    const deployedIndexId = await this.getDeployedIndexId();
+    const requestQuery: FindNeighborsRequestQuery = {
+      neighborCount: k,
+      datapoint: {
+        datapointId: `0`,
+        featureVector: query,
+      },
+    };
+    if (filter) {
+      requestQuery.datapoint.restricts = filter;
+    }
+    const request: FindNeighborsRequest = {
+      deployedIndexId,
+      queries: [requestQuery],
+    };
+
+    // Build the connection.
+    // Has to be done here, since we defer getting the endpoint until
+    // we need it.
+    const apiEndpoint = await this.getPublicAPIEndpoint();
+    const findNeighborsParams: FindNeighborsConnectionParams = {
+      endpoint: apiEndpoint,
+      indexEndpoint: this.indexEndpoint,
+      apiVersion: this.apiVersion,
+      location: this.location,
+      deployedIndexId,
+    };
+    const connection = new FindNeighborsConnection(
+      findNeighborsParams,
+      this.caller
+    );
+
+    // Make the call
+    const options = {};
+    const response = await connection.request(request, options);
+
+    // Get the document for each datapoint id and return them
+    const nearestNeighbors = response?.data?.nearestNeighbors ?? [];
+    const nearestNeighbor = nearestNeighbors[0];
+    const neighbors = nearestNeighbor?.neighbors ?? [];
+    const ret: [Document, number][] = await Promise.all(
+      neighbors.map(async (neighbor) => {
+        const id = neighbor?.datapoint?.datapointId;
+        const distance = neighbor?.distance;
+        let doc: IdDocument;
+        try {
+          doc = await this.docstore.search(id);
+        } catch (xx) {
+          console.error(xx);
+          doc = new Document({ pageContent: `Missing document ${id}` });
```

Comment:
I think it'd be better to return `undefined` or `null` and filter it out at the end - otherwise it'd cause problems downstream in retrieval.

Reply:
I go back and forth on that. What is the best way to handle a document that is indexed, but can't be found in the store? I went this way when I was using InMemoryStore and it was getting cleared out between tests, for obvious reasons, but the Matching Engine still had the index.

Are there cases where we want to know (and can benefit from) a document is "missing"?

The biggest downside I see is that "k-nearest neighbor" should return k documents, but filtering them out in this edge cases becomes deceptive.

However... I'm not sure I have a strong opinion to keep it this way. It is certainly an edge cases and not one I'd worry too much about, except that it should log the error.

Reply:
I changed my mind - I'm now more firmly of the opinion that they need to stay in, or there needs to be an option that keeps them in, or we need some way to register when they don't exist.

If we remove them, then there is no programatic way through LangChain JS to get "phantom documents" that are indexed, but not in the document store. If there was a _callbacks parameter where we could register this, I might consider it and have some kind of callback to indicate what phantom documents were removed... but this isn't implemented on `similaritySearchVectorWithScore()`, and even if it was, I'm not sure we have a callback event that can handle this.

Reply:
Ah right. Don't really see a great solution here... too bad the matching engine doesn't store the docs as well.

Reply:
There are a lot of things I wish the Matching Engine did differently... but...